whelktool/src/main/groovy/whelk/datatool/form/MatchForm.groovy

```diff
@@ -0,0 +1,307 @@
+        return mappings
+    }
+
+    static List<String> dropIndexes(List path) {
+        return path.findAll { it instanceof String } as List<String>
+    }
+
+    String getSparqlPattern(Map context) {
+        Map thing = getSparqlPreparedForm()
+        Map record = (Map) thing.remove(RECORD_KEY) ?: [:]
+
+        record[ID_KEY] = getRecordTmpId()
+        thing[ID_KEY] = getThingTmpId()
+        record[THING_KEY] = [(ID_KEY): getThingTmpId()]
+
+        Map graph = [(GRAPH_KEY): [record, thing]]
+
+        String ttl = toTurtleData(graph, context)
+
+        return insertTypeMappings(insertIdMappings(insertVars(ttl)))
+    }
+
+    private Map getSparqlPreparedForm() {
+        Map matchFormCopy = (Map) Document.deepCopy(form)
+
+        DocumentUtil.traverse(matchFormCopy) { node, path ->
+            if (node instanceof Map) {
+                def bNodeId = node.remove(BNODE_ID)
+                if (!bNodeId) return
+                node.remove(HAS_ID)
+                if (node[TYPE_KEY] == ANY_TYPE) {
+                    node.remove(TYPE_KEY)
+                }
+                if (asList(node.remove(MATCHING_MODE)).contains(SUBTYPES)) {
+                    def baseType = node.remove(TYPE_KEY)
+                    node[HAS_BASE_TYPE_TMP] = baseType
+                }
+                if (nodeIdMappings.containsKey(bNodeId)) {
+                    node[ID_KEY] = bNodeId
+                }
+                return new DocumentUtil.Nop()
+            }
+            if (asList(node).isEmpty()) {
+                return new DocumentUtil.Replace([:])
+            }
+        }
+
+        return matchFormCopy
+    }
+
+    private String insertVars(String ttl) {
+        def substitutions = [
+                ("<" + getThingTmpId() + ">") : getVar(getThingTmpId()),
+                ("<" + getRecordTmpId() + ">"): getVar(getRecordTmpId())
+        ]
+
+        baseTypeMappings.keySet().each { baseType ->
+            substitutions.put(":$HAS_BASE_TYPE_TMP \"$baseType\"".toString(), "a ?" + baseType)
+        }
+
+        nodeIdMappings.keySet().each { _id ->
+            substitutions.put("<" + _id + ">", getVar(_id))
+        }
+
+        return ttl.replace(substitutions)
+    }
+
+
+    private String insertTypeMappings(String sparqlPattern) {
+        def valuesClause = baseTypeMappings.collect { baseType, subTypes ->
+            "VALUES ?$baseType { ${([baseType] + subTypes).collect { ":$it" }.join(" ")} }\n"
+        }.join()
+        return valuesClause + sparqlPattern
+    }
+
+    private String insertIdMappings(String sparqlPattern) {
+        def valuesClauses = nodeIdMappings.collect { _id, ids ->
+            "VALUES ${getVar(_id)} { ${ids.collect { "<$it>" }.join(" ")} }\n"
+        }.join()
+        return valuesClauses + sparqlPattern
+    }
+
+    String getVar(String bNodeId) {
+        return bNodeId == getRecordTmpId()
+                ? "?$GRAPH_VAR"
+                : "?${bNodeId.replace('#', '')}"
+    }
+
+    @Memoized
```

Comment:
We have to watch out here if we ever want to implement "hot reloading" of vocab.
Better to put `@Memoized` on `jsonLd.getSubClasses(type)`?

Reply:
What's the benefit? Wouldn't that mean we risk getting outdated vocab data in even more places?

Reply:
I don't think so.
a JsonLd instance contains one "version" of vocab+context+display. 
For now it is never/updated replaced in one instance of whelk.
Keeping all the code that caches calculations made on vocab in one place reduces the risk of missing something if we ever change that.

Reply:
Good point. I'll move `@Memoized` to the "source".

Reply:
Ok!
Let's do it on develop though and not on the release branch.
Such a change in the core should have a bit more time for testing.

Reply:
Was thinking the same :+1: 

9be79a03301f1974d125a784205f3272abadd9ae